File: prime.py
# NOTICE!!! This is copied from https://stackoverflow.com/questions/4643647/fast-prime-factorization-module

import random

def primesbelow(N):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    #""" Input N>=6, Returns a list of primes, 2 <= p < N """
    correction = N % 6 > 1
    N = {0:N, 1:N-1, 2:N+4, 3:N+3, 4:N+2, 5:N+1}[N%6]
    sieve = [True] * (N // 3)
    sieve[0] = False
    for i in range(int(N ** .5) // 3 + 1):
        if sieve[i]:
            k = (3 * i + 1) | 1
            sieve[k*k // 3::2*k] = [False] * ((N//6 - (k*k)//6 - 1)//k + 1)
            sieve[(k*k + 4*k - 2*k*(i%2)) // 3::2*k] = [False] * ((N // 6 - (k*k + 4*k - 2*k*(i%2))//6 - 1) // k + 1)
    return [2, 3] + [(3 * i + 1) | 1 for i in range(1, N//3 - correction) if sieve[i]]

smallprimeset = set(primesbelow(100000))
_smallprimeset = 100000

def isprime(n, precision=7):
    # http://en.wikipedia.org/wiki/Miller-Rabin_primality_test#Algorithm_and_running_time
    if n == 1 or n % 2 == 0:
        return False
    elif n < 1:
        raise ValueError("Out of bounds, first argument must be > 0")
    elif n < _smallprimeset:
        return n in smallprimeset


    d = n - 1
    s = 0
    while d % 2 == 0:
        d //= 2
        s += 1

    for repeat in range(precision):
        a = random.randrange(2, n - 2)
        x = pow(a, d, n)

        if x == 1 or x == n - 1: continue

        for r in range(s - 1):
            x = pow(x, 2, n)
            if x == 1: return False
            if x == n - 1: break
        else: return False

    return True

# https://comeoncodeon.wordpress.com/2010/09/18/pollard-rho-brent-integer-factorization/
def pollard_brent(n):
    if n % 2 == 0: return 2
    if n % 3 == 0: return 3

    y, c, m = random.randint(1, n-1), random.randint(1, n-1), random.randint(1, n-1)
    g, r, q = 1, 1, 1
    while g == 1:
        x = y
        for i in range(r):
            y = (pow(y, 2, n) + c) % n

        k = 0
        while k < r and g==1:
            ys = y
            for i in range(min(m, r-k)):
                y = (pow(y, 2, n) + c) % n
                q = q * abs(x-y) % n
            g = gcd(q, n)
            k += m
        r *= 2
    if g == n:
        while True:
            ys = (pow(ys, 2, n) + c) % n
            g = gcd(abs(x - ys), n)
            if g > 1:
                break

    return g

smallprimes = primesbelow(10000) # might seem low, but 1000*1000 = 1000000, so this will fully factor every composite < 1000000
def primefactors(n, sort=False):
    factors = []

    limit = int(n ** .5) + 1
    for checker in smallprimes:
        if checker > limit: break
        while n % checker == 0:
            factors.append(checker)
            n //= checker
            limit = int(n ** .5) + 1
            if checker > limit: break

    if n < 2: return factors

    while n > 1:
        if isprime(n):
            factors.append(n)
            break
        factor = pollard_brent(n) # trial division did not fully factor, switch to pollard-brent
        factors.extend(primefactors(factor)) # recurse to factor the not necessarily prime factor returned by pollard-brent
        n //= factor

    if sort: factors.sort()

    return factors

def factorization(n):
    factors = {}
    for p1 in primefactors(n):
        try:
            factors[p1] += 1
        except KeyError:
            factors[p1] = 1
    return factors

totients = {}
def totient(n):
    if n == 0: return 1

    try: return totients[n]
    except KeyError: pass

    tot = 1
    for p, exp in factorization(n).items():
        tot *= (p - 1)  *  p ** (exp - 1)

    totients[n] = tot
    return tot

def gcd(a, b):
    if a == b: return a
    while b > 0: a, b = b, a % b
    return a

def lcm(a, b):
    return abs(a * b) // gcd(a, b)

File: prime.php
<?php

set_include_path(get_include_path().PATH_SEPARATOR.dirname(__FILE__).DIRECTORY_SEPARATOR.'libpy2php');
require_once 'libpy2php.php';



class PrimeModule
{
    public function __construct()
    {
        $this->smallprimeset = array_unique($this->primesbelow(100000));
        $this->_smallprimeset = 100000;
        $this->smallprimes = $this->primesbelow(10000);
    }

    public function primesbelow($N)
    {
        $res = [];
        for ($i = 2; $i <= $N; $i++) {
            if ($i % 2 != 1 && $i != 2) {
                continue;
            }
            $d = 3;
            $x = sqrt($i);
            while ($i % $d != 0 && $d < $x) {
                $d += 2;
            }
            if ((($i % $d == 0 && $i != $d) * 1) == 0) {
                $res[] = $i;
            }
        }

        return $res;
    }

    public function isprime($n, $precision = 7)
    {
        if (($n == 1) || (($n % 2) == 0)) {
            return false;
        } elseif (($n < 1)) {
            throw new Exception('Out of bounds, first argument must be > 0');
        } elseif (($n < $this->_smallprimeset)) {
            return in_array($n, $this->smallprimeset);
        }
        $d = ($n - 1);
        $s = 0;
        while (($d % 2) == 0) {
            $d = floor($d / 2);
            $s++;
        }
        $break = false;
        foreach (pyjslib_range($precision) as $repeat) {
            $a = rand(2, ($n - 2));
            $x = posmod(pow($a, $d), $n);
            if (($x == 1) || ($x == ($n - 1))) {
                continue;
            }
            foreach (pyjslib_range($s - 1) as $r) {
                $x = posmod(pow($x, 2), $n);
                if (($x == 1)) {
                    return false;
                }
                if (($x == ($n - 1))) {
                    $break = true;
                }
            }
            if (!$break) {
                return false;
            }
        }

        return true;
    }
    // taken from https://github.com/enricostara/telegram-mt-node/blob/master/lib/security/pq-finder.js
    public function getpq($pq) {
        $zero = new \phpseclib\Math\BigInteger(0);
        $one = new \phpseclib\Math\BigInteger(1);
        $two = new \phpseclib\Math\BigInteger(2);
        $three = new \phpseclib\Math\BigInteger(3);

        $p = new \phpseclib\Math\BigInteger();
        $q = new \phpseclib\Math\BigInteger();
	while (!$pq->equals($p->multiply($q))) {
            for ($i = 0; $i < 3; $i++) {
                $q = new \phpseclib\Math\BigInteger((random_int(0, 128) & 15) + 17);
                $x = new \phpseclib\Math\BigInteger(random_int(0, 1000000000) + 1);
                $y = $x;
                $lim = 1 << ($i + 18);
                for ($j = 1; $j < $lim; $j++) {
                    $a = $x;
                    $b = $x;
                    $c = $q;
                    while (!$b->equals($zero)) {
                        if ($b->powMod($one, $two)->equals($zero)) {
                           $c = $c->add($a);
                           if ($c->compare($pq) > 0) {
                                $c = $c->subtract($pq);
                           }
                        }
                        $a = $a->add($a);
                        if ($a->compare($pq) > 0) {
                            $a = $a->subtract($pq);
                        }
                        $b = $b->rightShift(1);
                    }
                    $x = $c;
                    $z = ($y->compare($x) > 0) ? $y->subtract($x) : $x->subtract($y);
                    $p = $z->gcd($pq);
                    if (!$p->equals($one)) {
                       break;
                    }
                    if (($j & ($j - 1)) === 0) {
                        $y = $x;
                    }
                }
                if (prime.gt(BigInteger.One())) {
                    break;
                }
            }
            $q = $pq->divide(prime)[0];
        }
        $_pq = ($q->compare($p) > 0) ? [$p, $q] : [$q, $p];
        return $_pq;
    }
    public function pollard_brent($n)
    {
        $zero = new \phpseclib\Math\BigInteger(0);
        $one = new \phpseclib\Math\BigInteger(1);
        $two = new \phpseclib\Math\BigInteger(2);
        $three = new \phpseclib\Math\BigInteger(3);
        if ($n->powMod($one, $two)->toString() == '0') {
            return 2;
        }
        if ($n->powMod($one, $three)->toString() == '0') {
            return 3;
        }
        $big = new \phpseclib\Math\BigInteger();
        $max = $n->subtract($one);
        list($y, $c, $m) = [new \phpseclib\Math\BigInteger(87552211475113995), new \phpseclib\Math\BigInteger(330422027228888537), new \phpseclib\Math\BigInteger(226866727920975483)];
        //[$big->random($one, $max), $big->random($one, $max), $big->random($one, $max)];
        list($g, $r, $q) = [$one, $one, $one];
        while ($g->equals($one)) {
            $x = $y;
            $params = ['y' => $y, 'two' => $two, 'c' => $c, 'one' => $one, 'n' => $n];
            $r->loopforeach(function ($i, $params) {
                $params['y'] = $params['y']->powMod($params['two'], $params['n'])->add($params['c'])->powMod($params['one'], $params['n']);
            }, $params);
            each($params);
            $k = $zero;
            while ($k->compare($r) == -1 && $g->equals($one)) {
                $ys = $y;
                $params = ['x' => $x, 'y' => $y, 'two' => $two, 'c' => $c, 'one' => $one, 'n' => $n, 'q' => $q];
                $m->min($r->subtract($k))->loopforeach(function ($i, $params) {
                    $params['y'] = $params['y']->powMod($params['two'], $params['n'])->add($params['c'])->powMod($params['one'], $params['n']);
                    $params['q'] = $params['q']->multiply($params['x']->subtract($params['y'])->abs())->powMod($params['one'], $params['n']);
                }, $params);
                each($params);
                $g = $q->gcd($n);
                $k = $k->add($m);
            }
            $r = $r->multiply($two);
        }
        die;
        if ($g->equals($n)) {
            while (true) {
                $ys = $ys->powMod($two, $n)->add($c)->powMod($one, $n);
                $g = $x->subtract($ys)->abs()->gcd($n);
                if ($g->compare($one) == 1) {
                    break;
                }
            }
        }

        return $g;
    }

    public function primefactors($n, $sort = false)
    {
        $factors = [];
        $n = new \phpseclib\Math\BigInteger(1724114033281923457);
var_dump($this->getpq($n));
        $one = new \phpseclib\Math\BigInteger(1);
        $two = new \phpseclib\Math\BigInteger(2);
        $limit = $n->root()->add($one);
        foreach ($this->smallprimes as $checker) {
            $checker = new \phpseclib\Math\BigInteger($checker);
            if ($limit->compare($checker) == -1) {
                break;
            }
            while ($n->modPow($one, $checker)->toString() == '0') {
                $factors[] = $checker;
                $n = $n->divide($checker)[0];
                $limit = $n->root()->add($one);
                if ($limit->compare($checker) == -1) {
                    break;
                }
            }
        }
        if ($n->compare($two) == -1) {
            return $factors;
        }
        while ($n->compare($two) == 1) {
            if ($n->isprime()) {
                $factors[] = $n;
                break;
            }
            $factor = $this->pollard_brent($n);
            $factors[] = $this->primefactors($factor);
            $n = floor($n / $factor);
        }
        if ($sort) {
            $factors = sort($factors);
        }

        return $factors;
    }

    public function factorization($n)
    {
        $factors = [];
        foreach (primefactors($n) as $p1) {
            if (isset($factors[$p1])) {
                $factors[$p1] += 1;
            } else {
                $factors[$p1] = 1;
            }
        }

        return $factors;
    }

    public function totient($n)
    {
        $totients = [];
        if (($n == 0)) {
            return 1;
        }
        if (isset($totients[$n])) {
            return $totients[$n];
        }
        $tot = 1;
        foreach (factorization($n) as $p => $exp) {
            $tot *= (($p - 1) * pow($p, ($exp - 1)));
        }
        $totients[$n] = $tot;

        return $tot;
    }

    public function gcd($a, $b)
    {
        if (($a == $b)) {
            return $a;
        }
        while (($b > 0)) {
            list($a, $b) = [$b, posmod($a, $b)];
        }

        return $a;
    }

    public function lcm($a, $b)
    {
        return floor(abs(($a * $b)) / $this->gcd($a, $b));
    }

}
